### PR TITLE
Timeshift

### PR DESCRIPTION
**Pull request by phunkyfish**
v3.6.0
- Added: Support for Timeshifting

As the M3U8 playlist needs to updated on a rolling basis it was not possible to just direct the stream into kodi. Instead download each segment/decrypt etc and then stream into kodi.

Leveraged a C project for the HLS parts (most of it ported): https://github.com/e2iplayer/hlsdl and a concurrent queue for the stream: https://github.com/cameron314/readerwriterqueue

### CODE REVIEW

**Comment by afedchin** [comment]
who will support all this code?
my personal as an author of the addon opinion: if you need timeshift or recordings then IPTV addon is not your choice.

**Comment by phunkyfish** [comment]
I'll agree it turned into a bit of a beast. I have no issue with supporting it. It was mentioned on the forum that timeshift was acceptable here but recordings would not be. I thought the community would appreciate it, plus I had some time spare time over the Xmas break. The addon is unchanged unless you enable timeshift.

Ideally something like inputstream.adaptive could handle this but I wasn't sure where to start with that.

**Comment by phunkyfish** [comment]
Strange rapidxml appears to be breaking the build. @Rechi thoughts? Or have I just done something wrong.

**Comment by Rechi** [comment]
#216 fixes the build errors.

**Comment by phunkyfish** [comment]
Great, thank you

File: CMakeLists.txt [hunk 1]
```diff
@@ -11,23 +11,65 @@ find_package(kodiplatform REQUIRED)
 find_package(p8-platform REQUIRED)
 find_package(Rapidxml REQUIRED)
 find_package(ZLIB REQUIRED)
+find_package(OpenSSL REQUIRED)
```
Comment (phunkyfish) [inline CMakeLists.txt:14]: @Rechi I used vfs.sftp and vfs.libarchive as a reference but can't see why this fails on nearly every platform. Am I missing something?

Comment (Rechi) [reply CMakeLists.txt:14]: OpenSSL is missing in depends.

Comment (phunkyfish) [reply CMakeLists.txt:14]: On Jenkins?

Comment (Rechi) [reply CMakeLists.txt:14]: in your PR branch

Comment (phunkyfish) [reply CMakeLists.txt:14]: Oh, now I feel silly. Thanks again for your help.

File: debian/control [hunk 2]
```diff
@@ -2,7 +2,7 @@ Source: kodi-pvr-iptvsimple
 Priority: extra
 Maintainer: Anton Fedchin <anightik@gmail.com>
 Build-Depends: debhelper (>= 9.0.0), cmake, libkodiplatform-dev (>= 16.0.0),
-               kodi-addon-dev, librapidxml-dev, zlib1g-dev
+               kodi-addon-dev, librapidxml-dev, zlib1g-dev, openssl
```
Comment (Rechi) [inline debian/control]: openssl is the openssl binary, you need the openssl library and headers (libssl-dev)

Comment (phunkyfish) [reply debian/control]: Ah, ok, thanks

**Comment by ksooo** [comment]
> who will support all this code?
my personal as an author of the addon opinion: if you need timeshift or recordings then IPTV addon is not your choice.

@phunkyfish reading this from the addon maintainer I'm not sure it is worth to invest more time into this effort.

**Comment by phunkyfish** [comment]
I just wanted to learn how to get it working. A personal learning exercise 😉

**Comment by SerhiyRomanov** [comment]
Hello @phunkyfish 

> I thought the community would appreciate it

Yes, I appreciate your work because I still don't understand why the timeshift feature doesn't work here. 
My IPTV operator provides an archive for several days and I am very happy about Catchup (Archive) on Matrix, additional thanks for it.

**Comment by phunkyfish** [comment]
I do hope to support timeshift one day.

The new inputstream addon ffmpegdirect that is used for catchup could potentially support it in the future.

**Comment by phunkyfish** [comment]
Closing, timeshift feature released on Matrix using inputstream.ffmpegdirect.

**Pull request closed without merging**